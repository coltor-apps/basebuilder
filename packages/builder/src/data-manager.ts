import {
  createSubscriptionManager,
  type Subscribe,
} from "./subscription-manager";

export function createDataManager<TData>(initialData: TData): {
  getData: () => TData;
  setData: (data: TData) => TData;
  subscribe: Subscribe<TData>;
} {
  let data: TData = initialData;

  function getData(): TData {
    return data;
  }

  const { notify, subscribe } = createSubscriptionManager<TData>();

  return {
    subscribe,
    getData,
    setData(newData) {
      data = newData;

      notify(data);

      return data;
    },
  };
}
