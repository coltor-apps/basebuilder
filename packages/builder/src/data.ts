import { createSubscriptionManager, type Subscribe } from "./subscription";

export function createDataManager<TData>(data: TData): {
  getData: () => TData;
  setData: (setter: (oldData: TData) => TData) => TData;
  subscribe: Subscribe<TData>;
} {
  let localData: TData = data;

  function getData(): TData {
    return localData;
  }

  const { notify, subscribe } = createSubscriptionManager<TData>();

  return {
    subscribe,
    getData,
    setData(setter) {
      localData = setter(getData());

      notify(localData);

      return localData;
    },
  };
}
