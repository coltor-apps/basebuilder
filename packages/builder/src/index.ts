export { createBuilder } from "./builder";
export type { Builder, BuilderEntities } from "./builder";

export { createEntity } from "./entity";
export type { Entity } from "./entity";

export { createInput } from "./input";
export type { Input } from "./input";

export { SchemaValidationError as EntityValidationError, createStore } from "./store";
export type {
  Store,
  StoreData,
  StoreEntity,
  SchemaValidationErrorCode as EntityValidationErrorCause,
} from "./store";

export { getEntitiesNamesExcept } from "./utils";
