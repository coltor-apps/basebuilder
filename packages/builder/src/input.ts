export interface Input<TName extends string, TValue> {
  name: TName;
  validate: (value: unknown) => TValue;
  defaultValue: () => Awaited<TValue> | undefined;
}

export type InputsValues<
  TInputs extends ReadonlyArray<Input<string, unknown>>,
> = {
  [K in TInputs[number]["name"]]: Awaited<
    ReturnType<Extract<TInputs[number], { name: K }>["validate"]>
  >;
};

type OptionalInputArgs = "defaultValue";

export function createInput<const TName extends string, TValue>(
  options: Omit<Input<TName, TValue>, OptionalInputArgs> &
    Partial<Pick<Input<TName, TValue>, OptionalInputArgs>>,
): Input<TName, TValue> {
  function fallbackDefaultValue() {
    return undefined;
  }

  return {
    ...options,
    defaultValue: options.defaultValue ?? fallbackDefaultValue,
  };
}
