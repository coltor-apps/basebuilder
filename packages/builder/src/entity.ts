import { type Input, type InputsValues } from "./input";

export interface EntityContext<TInputs extends ReadonlyArray<Input<string>>> {
  inputs: InputsValues<TInputs>;
}

export interface Entity<
  TName extends string,
  TInputs extends ReadonlyArray<Input<string>>,
  TValue,
> {
  name: TName;
  validate: (
    value: unknown,
    context: EntityContext<TInputs>,
  ) => Promise<TValue> | TValue;
  defaultValue: (context: EntityContext<TInputs>) => TValue | undefined;
  inputs: TInputs;
}

type OptionalEntityArgs = "inputs" | "validate" | "defaultValue";

export function createEntity<
  const TName extends string,
  const TInputs extends ReadonlyArray<Input<string>>,
  TValue,
>(
  options: Omit<Entity<TName, TInputs, TValue>, OptionalEntityArgs> &
    Partial<Pick<Entity<TName, TInputs, TValue>, OptionalEntityArgs>>,
): Entity<TName, TInputs, TValue> {
  return {
    ...options,
    validate:
      options.validate ??
      ((value) => {
        if (value) {
          throw new Error("No value allowed");
        }

        return undefined as TValue;
      }),
    inputs: options.inputs ?? ([] as ReadonlyArray<unknown> as TInputs),
    defaultValue: options.defaultValue ?? (() => undefined),
  };
}
