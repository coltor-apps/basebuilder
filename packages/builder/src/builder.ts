import { type Entity } from "./entity";
import { type Input } from "./input";
import { generateUuid, validateUuid } from "./uuid";

export type BuilderEntities = ReadonlyArray<
  Entity<string, ReadonlyArray<Input<string, unknown>>, unknown>
>;

type ChildrenAllowed<TEntities extends BuilderEntities> = {
  [K in TEntities[number]["name"]]?:
    | ReadonlyArray<TEntities[number]["name"]>
    | true;
};

export type Builder<
  TEntities extends BuilderEntities = BuilderEntities,
  TChildrenAllowed extends ChildrenAllowed<TEntities> = Record<
    string,
    true | ReadonlyArray<string> | undefined
  >,
  TParentRequired extends ReadonlyArray<
    TEntities[number]["name"]
  > = ReadonlyArray<string>,
> = {
  entities: TEntities;
  entityId: {
    generate: () => string;
    validate: (id: string) => void;
  };
  childrenAllowed: TChildrenAllowed;
  parentRequired: TParentRequired;
};

type OptionalBuilderArgs = "childrenAllowed" | "parentRequired";

export function createBuilder<
  const TEntities extends BuilderEntities,
  const TChildrenAllowed extends ChildrenAllowed<TEntities> = Record<
    string,
    never
  >,
  const TParentRequired extends ReadonlyArray<TEntities[number]["name"]> = [],
>(
  options: Omit<
    Builder<TEntities, TChildrenAllowed, TParentRequired>,
    OptionalBuilderArgs | "entityId"
  > &
    Partial<
      Pick<
        Builder<TEntities, TChildrenAllowed, TParentRequired>,
        OptionalBuilderArgs
      >
    > & {
      entityId?: Partial<
        Builder<TEntities, TChildrenAllowed, TParentRequired>["entityId"]
      >;
    },
): Builder<TEntities, TChildrenAllowed, TParentRequired> {
  return {
    ...options,
    entityId: {
      generate: options.entityId?.generate ?? generateUuid,
      validate: options.entityId?.validate ?? validateUuid,
    },
    childrenAllowed: options.childrenAllowed ?? ({} as TChildrenAllowed),
    parentRequired:
      options.parentRequired ??
      ([] as ReadonlyArray<unknown> as TParentRequired),
  };
}

export function getEntityDefinition(
  entityType: string,
  builder: Builder,
): Builder["entities"][number] | undefined {
  return builder.entities.find(
    (builderEntity) => builderEntity.name === entityType,
  );
}

export function ensureEntityIsRegistered(
  entityType: string,
  builder: Builder,
): Builder["entities"][number] {
  const entityDefinition = getEntityDefinition(entityType, builder);

  if (!entityDefinition) {
    throw new Error(`Unkown entity type "${entityType}".`);
  }

  return entityDefinition;
}

export function ensureEntityInputIsRegistered(
  entityType: string,
  inputName: string,
  builder: Builder,
): Input {
  const entityDefinition = ensureEntityIsRegistered(entityType, builder);

  const input = entityDefinition.inputs.find(
    (input) => input.name === inputName,
  );

  if (!input) {
    throw new Error(`Unkown entity input "${inputName}".`);
  }

  return input;
}

export function ensureEntityInputsAreRegistered(
  entityType: string,
  inputNames: Array<string>,
  builder: Builder,
): Array<Input> {
  const inputs = inputNames.map((inputName) =>
    ensureEntityInputIsRegistered(entityType, inputName, builder),
  );

  return inputs;
}

export function isEntityChildAllowed(
  entityType: string,
  childEntityType: string,
  builder: Builder,
): boolean {
  const allowedChildren = builder.childrenAllowed[entityType];

  if (!allowedChildren) {
    return false;
  }

  return allowedChildren === true || allowedChildren.includes(childEntityType);
}

export function isEntityParentRequired(
  entityType: string,
  builder: Builder,
): boolean {
  return builder.parentRequired.includes(entityType);
}

export function ensureEntityChildAllowed(
  entityType: string,
  childEntityType: string,
  builder: Builder,
): void {
  if (!isEntityChildAllowed(entityType, childEntityType, builder)) {
    throw new Error("Child is not allowed.");
  }
}

export function ensureEntityCanLackParent(
  entityType: string,
  builder: Builder,
): void {
  if (isEntityParentRequired(entityType, builder)) {
    throw new Error("A parent is required.");
  }
}
