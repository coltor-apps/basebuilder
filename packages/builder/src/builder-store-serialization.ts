import { type Builder } from "./builder";
import {
  type BuilderStoreData,
  type BuilderStoreEntity,
  type SerializedBuilderStoreData,
} from "./builder-store";

export function serializeBuilderStoreData<TBuilder extends Builder>(
  data: BuilderStoreData<TBuilder>,
): SerializedBuilderStoreData<TBuilder> {
  const newEntities: SerializedBuilderStoreData<TBuilder>["schema"]["entities"] =
    {};

  for (const [id, entity] of data.schema.entities) {
    const { children, ...entityData } = entity;

    newEntities[id] = {
      ...entityData,
      ...(children ? { children: Array.from(children) } : {}),
      inputs:
        entityData.inputs as unknown as SerializedBuilderStoreData<TBuilder>["schema"]["entities"][string]["inputs"],
    };
  }

  return {
    schema: {
      root: Array.from(data.schema.root),
      entities: newEntities,
    },
    entitiesInputsErrors: Object.fromEntries(data.entitiesInputsErrors),
  };
}

export function deserializeBuilderStoreData<TBuilder extends Builder>(
  data: SerializedBuilderStoreData<TBuilder>,
): BuilderStoreData<TBuilder> {
  return {
    schema: {
      entities: new Map(
        Object.entries(data.schema.entities).map(([id, entity]) => [
          id,
          {
            ...entity,
            ...(entity.children ? { children: new Set(entity.children) } : {}),
            inputs:
              entity.inputs as unknown as BuilderStoreEntity<TBuilder>["inputs"],
          },
        ]),
      ),
      root: new Set(data.schema.root),
    },
    entitiesInputsErrors: new Map(Object.entries(data.entitiesInputsErrors)),
  };
}
