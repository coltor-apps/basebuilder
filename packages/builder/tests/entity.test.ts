import { describe, expect, it } from "vitest";
import { z } from "zod";

import { createEntity } from "../src/entity";
import { createInput } from "../src/input";

describe("entity", () => {
  it("can be created with minimal options", () => {
    const entity = createEntity({
      name: "text",
    });

    expect(entity).toMatchSnapshot();

    expect(entity.defaultValue({ inputs: {}, values: {} })).toMatchSnapshot();
  });

  it("can validate values", () => {
    const entity = createEntity({
      name: "text",
      validate(value) {
        return z.string().parse(value);
      },
    });

    expect(
      entity.validate("valid", { inputs: {}, values: {} }),
    ).toMatchSnapshot();

    expect(() =>
      entity.validate(1, { inputs: {}, values: {} }),
    ).toThrowErrorMatchingSnapshot();
  });

  it("throws when validating without a validator", () => {
    const entity = createEntity({
      name: "text",
    });

    expect(() =>
      entity.validate("value", { inputs: {}, values: {} }),
    ).toThrowErrorMatchingSnapshot();

    expect(
      entity.validate(undefined, { inputs: {}, values: {} }),
    ).toMatchSnapshot();
  });

  it("can be created with default value", () => {
    const entity = createEntity({
      name: "text",
      defaultValue() {
        return "test";
      },
    });

    expect(entity.defaultValue({ inputs: {}, values: {} })).toMatchSnapshot();
  });

  it("can be created with inputs", () => {
    const entity = createEntity({
      name: "text",
      inputs: [
        createInput({
          name: "required",
          validate(value) {
            return value;
          },
        }),
      ],
      validate(value) {
        return value;
      },
    });

    expect(entity).toMatchSnapshot();
  });
});
