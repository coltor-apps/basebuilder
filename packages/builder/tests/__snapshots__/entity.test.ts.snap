// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`entity > can be created with default value 1`] = `"test"`;

exports[`entity > can be created with inputs 1`] = `
{
  "defaultValue": [Function],
  "inputs": [
    {
      "defaultValue": [Function],
      "name": "required",
      "validate": [Function],
    },
  ],
  "isValueAllowed": true,
  "name": "text",
  "shouldBeProcessed": [Function],
  "validate": [Function],
}
`;

exports[`entity > can be created with minimal options 1`] = `
{
  "defaultValue": [Function],
  "inputs": [],
  "isValueAllowed": false,
  "name": "text",
  "shouldBeProcessed": [Function],
  "validate": [Function],
}
`;

exports[`entity > can be created with minimal options 2`] = `undefined`;

exports[`entity > can validate values 1`] = `"valid"`;

exports[`entity > can validate values 2`] = `
"[
  {
    \\"code\\": \\"invalid_type\\",
    \\"expected\\": \\"string\\",
    \\"received\\": \\"number\\",
    \\"path\\": [],
    \\"message\\": \\"Expected string, received number\\"
  }
]"
`;

exports[`entity > throws when validating without a validator 1`] = `"Values for entities of type 'text' are not allowed."`;

exports[`entity > throws when validating without a validator 2`] = `undefined`;
